### PR TITLE
wideresnet || cifar10 || fast.ai + students

### PR DESCRIPTION
**Pull request by brettkoonce**

### CODE REVIEW

File: CIFAR10/train/fastai_pytorch.json [hunk 1]
```diff
@@ -0,0 +1,11 @@
+{
+    "version": "v1.0",
+    "author": "fast.ai + students",
+    "authorEmail": "j@fast.ai",
+    "framework": "fastai / pytorch",
+    "codeURL": "https://github.com/fastai/imagenet-fast/",
+    "model": "Custom Wide Resnet",
+    "hardware": "8 * V100 (AWS p3.xlarge)",
+    "costPerHour": 8.68,
```
Comment (ddkang) [inline CIFAR10/train/fastai_pytorch.json]: We do not allow spot pricing. Please update the cost to reserved instance cost.

File: CIFAR10/train/fastai_pytorch.json [hunk 2]
```diff
@@ -0,0 +1,11 @@
+{
+    "version": "v1.0",
+    "author": "fast.ai + students",
+    "authorEmail": "j@fast.ai",
+    "framework": "fastai / pytorch",
+    "codeURL": "https://github.com/fastai/imagenet-fast/",
```
Comment (ddkang) [inline CIFAR10/train/fastai_pytorch.json]: Please add the commit hash for reproducibility purposes.

File: CIFAR10/train/fastai_pytorch.json [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+    "model": "Custom Wide Resnet",
```
Comment (ddkang) [inline CIFAR10/train/fastai_pytorch.json]: Could you expand on this a bit more? I believe wide resents are parameterized - what parameters do you use for your custom wide resnet?

File: CIFAR10/train/fastai_pytorch.tsv [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+40	0.05138888888888889	92.87
+41	0.05277777777777778	93.18
+42	0.05388888888888889	94.01
+43	0.05527777777777778	94.04
+44	0.05638888888888889	94.16
+45	0.0575	94.04
+46	0.058888888888888886	94.25
+47	0.06	94.23
+48	0.06111111111111111	94.26
+49	0.0625	94.21
```
Comment (ddkang) [inline CIFAR10/train/fastai_pytorch.tsv]: These timestamps do not seem very precise. Can you update your timer to at least millisecond precision?

**Review by ddkang — CHANGES_REQUESTED** [review]
Thank you for your submission! I've left some comments inline.

Additionally, we have a few questions for reproducibility purposes.

It seems you're using PyTorch to run the code. Can you provide the commit hash of PyTorch you used to run this?

The README says a fastai install needs to be symlinked into the directory. Could you provide step-by-step instructions, or a script to do this?

Could you also add a note about the learning rate schedule in the misc field?

File: CIFAR10/train/fastai_pytorch.json (hunk 1 above)
Comment (brettkoonce) [reply CIFAR10/train/fastai_pytorch.json]: updated

File: CIFAR10/train/fastai_pytorch.tsv (hunk 4 above)
Comment (brettkoonce) [reply CIFAR10/train/fastai_pytorch.tsv]: switch to millisecond timing and re-ran the script!

File: CIFAR10/train/fastai_pytorch.json (hunk 3 above)
Comment (brettkoonce) [reply CIFAR10/train/fastai_pytorch.json]: linked to the model we are using!

File: CIFAR10/train/fastai_pytorch.json (hunk 2 above)
Comment (brettkoonce) [reply CIFAR10/train/fastai_pytorch.json]: added the current master SHA1

**Comment by brettkoonce** [comment]
Put some more info in the readme as well:
   https://github.com/fastai/imagenet-fast/tree/master/cifar10/dawn_submission

**Comment by brettkoonce** [comment]
@ddkang Let me know if you need anything else!

**Comment by ddkang** [comment]
Great, thank you!

I still don't understand what you mean by "symlink a fastai install" - does this mean you'll need to symlink the git repository? The conda install location? Could you clarify what you mean here?

**Comment by brettkoonce** [comment]
@ddkang the calls like

    from fastai.transforms import *

fail unless you have the literal fastai/fastai subdirectory folder linked in your directory.  You should have something that looks like this when you run ls -l:

    lrwxrwxrwx 1 ubuntu ubuntu    23 Apr 13 20:46 fastai -> ../../../fastai/fastai/

pwd:

    /home/ubuntu/git/imagenet-fast/cifar10

Here is how to create a symlink (alias) on linux:

    https://www.lifewire.com/create-symbolic-links-ln-command-4059723
 
All we are doing is creating a folder inside the imagenet-fast cifar directory that allows us to load these functions from fastai's folder directly.

**Comment by codyaustun** [comment]
Looks good to me.

Side note: for the fastai dependency, I think you could just pip install it.

```
pip install git+https://github.com/fastai/fastai.git
```

File: CIFAR10/train/fastai_pytorch.json [hunk 5]
```diff
@@ -0,0 +1,17 @@
+{
+    "version": "v1.0",
+    "author": "fast.ai + students",
+    "authorEmail": "j@fast.ai",
+    "framework": "fastai / pytorch",
+    "codeURL": "https://github.com/fastai/imagenet-fast/commit/2748de3b0e28ecca1e02d90b2d5bf7d9b9e58e3d",
+    "model": "Custom Wide Resnet (https://github.com/fastai/imagenet-fast/blob/2748de3b0e28ecca1e02d90b2d5bf7d9b9e58e3d/cifar10/models/wideresnet.py#L60)",
```
Comment (codyaustun) [inline CIFAR10/train/fastai_pytorch.json]: @brettkoonce can you more the model link to `misc`? I don't think this will display well on the leaderboard.

**Review by codyaustun — CHANGES_REQUESTED** [review]

File: CIFAR10/train/fastai_pytorch.json (hunk 5 above)
Comment (brettkoonce) [reply CIFAR10/train/fastai_pytorch.json]: done!

Comment (codyaustun) [reply CIFAR10/train/fastai_pytorch.json]: Sorry I should have clarified. Change the `model` field to "Custom Wide ResNet" and move the url to `misc` maybe with a field called `modelURL`

**Review by codyaustun — APPROVED** [review]

File: CIFAR10/train/fastai_pytorch.json [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+    "author": "fast.ai + students",
+    "authorEmail": "j@fast.ai",
+    "framework": "fastai / pytorch",
+    "codeURL": "https://github.com/fastai/imagenet-fast/commit/2748de3b0e28ecca1e02d90b2d5bf7d9b9e58e3d",
+    "model": "Custom Wide Resnet",
+    "hardware": "8 * V100 (AWS p3.xlarge)",
+    "costPerHour": 24.48,
+    "timestamp": "2018-04-18",
+    "misc": {
+		"model implementation": "https://github.com/fastai/imagenet-fast/blob/2748de3b0e28ecca1e02d90b2d5bf7d9b9e58e3d/cifar10/models/wideresnet.py#L60",
```
Comment (ddkang) [inline CIFAR10/train/fastai_pytorch.json]: nit: mixed tab / spaces

File: CIFAR10/train/fastai_pytorch.json [hunk 7]
```diff
@@ -0,0 +1,18 @@
+{
+    "version": "v1.0",
+    "author": "fast.ai + students",
```
Comment (ddkang) [inline CIFAR10/train/fastai_pytorch.json]: nit: are there few enough students that you can credit them here?

**Review by ddkang — COMMENTED** [review]
Great, thanks! Aside from the nits, looks good

File: CIFAR10/train/fastai_pytorch.json (hunk 6 above)
Comment (brettkoonce) [reply CIFAR10/train/fastai_pytorch.json]: fixed!

File: CIFAR10/train/fastai_pytorch.json (hunk 7 above)
Comment (brettkoonce) [reply CIFAR10/train/fastai_pytorch.json]: added!

**Comment by deepakn94** [comment]
Great, thanks! Merged.

**Comment by deepakn94** [comment]
Looking at this again, looks like we have the instance type wrong -- should be a p3.16xlarge. Can we fix this in a separate PR? Thanks!
